### ISSUE TITLE
`inference_mode` before training results in FSDP AssertionError

### ISSUE THREAD

**Issue by carmocca** [pytorch/pytorch#113203]
### 🐛 Describe the bug

```python
import os
import torch
import torch.cuda
import torch.distributed as dist
import torch.multiprocessing as mp
from torch.distributed.fsdp import FullyShardedDataParallel as FSDP

def work(rank):
    os.environ["MASTER_ADDR"] = "127.0.0.1"
    os.environ["MASTER_PORT"] = "1234"
    dist.init_process_group("nccl", world_size=2, rank=rank)

    torch.cuda.set_device(rank)
    device = torch.device("cuda", rank)

    with device:
        model = torch.nn.Linear(10, 10)
        input_ids = torch.randn(2, 10)
    model = FSDP(model)

    with torch.inference_mode():
        print("Validating ...")
        model(input_ids)
    print("Training ...")
    model(input_ids)

if __name__ == "__main__":
    mp.spawn(work, nprocs=2)
```

The resulting error is:

```python
Validating ...
Validating ...
Training ...
The `grad_fn` is needed to access the `AccumulateGrad` and register the post-backward hook
Training ...
The `grad_fn` is needed to access the `AccumulateGrad` and register the post-backward hook
  File "<string>", line 1, in <module>
  File "/usr/lib/python3.10/multiprocessing/spawn.py", line 116, in spawn_main
    exitcode = _main(fd, parent_sentinel)
  File "/usr/lib/python3.10/multiprocessing/spawn.py", line 129, in _main
    return self._bootstrap(parent_sentinel)
  File "/usr/lib/python3.10/multiprocessing/process.py", line 314, in _bootstrap
    self.run()
  File "/usr/lib/python3.10/multiprocessing/process.py", line 108, in run
    self._target(*self._args, **self._kwargs)
  File "/home/carlos/nightly-env/lib/python3.10/site-packages/torch/multiprocessing/spawn.py", line 74, in _wrap
    fn(i, *args)
  File "/home/carlos/lit-parrot/repro.py", line 25, in work
    model(input_ids)
  File "/home/carlos/nightly-env/lib/python3.10/site-packages/torch/nn/modules/module.py", line 1519, in _wrapped_call_impl
    return self._call_impl(*args, **kwargs)
  File "/home/carlos/nightly-env/lib/python3.10/site-packages/torch/nn/modules/module.py", line 1528, in _call_impl
    return forward_call(*args, **kwargs)
  File "/home/carlos/nightly-env/lib/python3.10/site-packages/torch/distributed/fsdp/fully_sharded_data_parallel.py", line 849, in forward
    args, kwargs = _pre_forward(
  File "/home/carlos/nightly-env/lib/python3.10/site-packages/torch/distributed/fsdp/_runtime_utils.py", line 416, in _pre_forward
    _register_post_backward_hook(state, handle)
  File "/home/carlos/nightly-env/lib/python3.10/site-packages/torch/distributed/fsdp/_runtime_utils.py", line 1419, in _register_post_backward_hook
    _p_assert(
  File "/home/carlos/nightly-env/lib/python3.10/site-packages/torch/distributed/utils.py", line 145, in _p_assert
    traceback.print_stack()
  File "<string>", line 1, in <module>
  File "/usr/lib/python3.10/multiprocessing/spawn.py", line 116, in spawn_main
    exitcode = _main(fd, parent_sentinel)
  File "/usr/lib/python3.10/multiprocessing/spawn.py", line 129, in _main
    return self._bootstrap(parent_sentinel)
  File "/usr/lib/python3.10/multiprocessing/process.py", line 314, in _bootstrap
    self.run()
  File "/usr/lib/python3.10/multiprocessing/process.py", line 108, in run
    self._target(*self._args, **self._kwargs)
  File "/home/carlos/nightly-env/lib/python3.10/site-packages/torch/multiprocessing/spawn.py", line 74, in _wrap
    fn(i, *args)
  File "/home/carlos/lit-parrot/repro.py", line 25, in work
    model(input_ids)
  File "/home/carlos/nightly-env/lib/python3.10/site-packages/torch/nn/modules/module.py", line 1519, in _wrapped_call_impl
    return self._call_impl(*args, **kwargs)
  File "/home/carlos/nightly-env/lib/python3.10/site-packages/torch/nn/modules/module.py", line 1528, in _call_impl
    return forward_call(*args, **kwargs)
  File "/home/carlos/nightly-env/lib/python3.10/site-packages/torch/distributed/fsdp/fully_sharded_data_parallel.py", line 849, in forward
    args, kwargs = _pre_forward(
  File "/home/carlos/nightly-env/lib/python3.10/site-packages/torch/distributed/fsdp/_runtime_utils.py", line 416, in _pre_forward
    _register_post_backward_hook(state, handle)
  File "/home/carlos/nightly-env/lib/python3.10/site-packages/torch/distributed/fsdp/_runtime_utils.py", line 1419, in _register_post_backward_hook
    _p_assert(
  File "/home/carlos/nightly-env/lib/python3.10/site-packages/torch/distributed/utils.py", line 145, in _p_assert
    traceback.print_stack()
Traceback (most recent call last):
  File "/home/carlos/lit-parrot/repro.py", line 28, in <module>
    mp.spawn(work, nprocs=2)
  File "/home/carlos/nightly-env/lib/python3.10/site-packages/torch/multiprocessing/spawn.py", line 246, in spawn
    return start_processes(fn, args, nprocs, join, daemon, start_method="spawn")
  File "/home/carlos/nightly-env/lib/python3.10/site-packages/torch/multiprocessing/spawn.py", line 202, in start_processes
    while not context.join():
  File "/home/carlos/nightly-env/lib/python3.10/site-packages/torch/multiprocessing/spawn.py", line 163, in join
    raise ProcessRaisedException(msg, error_index, failed_process.pid)
torch.multiprocessing.spawn.ProcessRaisedException: 

-- Process 0 terminated with the following error:
Traceback (most recent call last):
  File "/home/carlos/nightly-env/lib/python3.10/site-packages/torch/multiprocessing/spawn.py", line 74, in _wrap
    fn(i, *args)
  File "/home/carlos/lit-parrot/repro.py", line 25, in work
    model(input_ids)
  File "/home/carlos/nightly-env/lib/python3.10/site-packages/torch/nn/modules/module.py", line 1519, in _wrapped_call_impl
    return self._call_impl(*args, **kwargs)
  File "/home/carlos/nightly-env/lib/python3.10/site-packages/torch/nn/modules/module.py", line 1528, in _call_impl
    return forward_call(*args, **kwargs)
  File "/home/carlos/nightly-env/lib/python3.10/site-packages/torch/distributed/fsdp/fully_sharded_data_parallel.py", line 849, in forward
    args, kwargs = _pre_forward(
  File "/home/carlos/nightly-env/lib/python3.10/site-packages/torch/distributed/fsdp/_runtime_utils.py", line 416, in _pre_forward
    _register_post_backward_hook(state, handle)
  File "/home/carlos/nightly-env/lib/python3.10/site-packages/torch/distributed/fsdp/_runtime_utils.py", line 1419, in _register_post_backward_hook
    _p_assert(
  File "/home/carlos/nightly-env/lib/python3.10/site-packages/torch/distributed/utils.py", line 147, in _p_assert
    raise AssertionError(s)
AssertionError: The `grad_fn` is needed to access the `AccumulateGrad` and register the post-backward hook
```

### Versions

`torch==2.2.0.dev20231102+cu121`

Initially reported in https://github.com/Lightning-AI/lit-gpt/issues/699

cc @zhaojuanmao @mrshenli @rohan-varma @awgu @fegin @penguinwu

**Comment by awgu** [comment]
It seems we should detect that we are in inference mode and skip registering the post-backward hook, augmenting the check here:
https://github.com/pytorch/pytorch/blob/356f3458c4dbe7dfc31019d0b54730dad977c24a/torch/distributed/fsdp/_runtime_utils.py#L1409-L1410

Would it be possible to try with `with torch.no_grad()` in addition to inference mode to unblock?

**Comment by carmocca** [comment]
Yes, I am using `no_grad` as a workaround.

**Comment by awgu** [comment]
Strange -- I was writing a unit test for this before making any changes:
```
@skip_if_lt_x_gpu(2)
def test_inference_mode(self):
    """
    Tests that FSDP works under inference mode and produces the same output
    as under no grad mode.
    """
    torch.manual_seed(42 + self.rank)
    with torch.device(self.rank):
        model = torch.nn.Linear(10, 10)
        inp = torch.randn(2, 10)
    model = FSDP(model)
    with torch.inference_mode():
        out_inference_mode = model(inp)
    with torch.no_grad():
        out_no_grad = model(inp)
    self.assertEqual(out_no_grad, out_inference_mode)
```
This works out of the box. I wonder why it is not erroring.

**Comment by awgu** [comment]
I realized this is pretty interesting. The error happens on the forward _not_ under `inference_mode()`. I wonder why there is no `grad_fn`.

**Comment by awgu** [comment]
I think this might be expected behavior in the sense that under inference mode, there is no autograd tracking, so the `FlatParameter` does not have a `grad_fn`. Since the root's `FlatParameter` is not freed after forward, it is preserved into the non-inference-mode forward, still without a `grad_fn`.

**Comment by carmocca** [comment]
Then shouldnt the flatparameters be reconstructed after the inference region is over? In the second forward

**Comment by awgu** [comment]
I think that the issue is that the root (or more generally when `reshard_after_forward == False` like for `SHARD_GRAD_OP`), the flat parameters are not reconstructed, but I would need to investigate more to confirm.

**Comment by mori360** [comment]
We are committed to support FSDP2. Here is the migration guide https://github.com/pytorch/torchtitan/blob/main/docs/fsdp.md
If you encounter any similar issues with FSDP2, please open a new issue with updated information.
